src/AutoRest.CSharp/Common/Output/PostProcessing/ReferenceMapBuilder.cs

```diff
@@ -0,0 +1,217 @@
+        {
+            if (keySymbol is not INamedTypeSymbol keyTypeSymbol)
+                return;
+            // add the class and all its partial classes to the map
+            // this will make all the partial classes are referencing each other in the reference map
+            // when we make the travesal over the reference map, we will not only remove one of the partial class, instead we will either keep all the partial classes (if at least one of them has references), or remove all of them (if none of them has references)
+            AddToReferenceMap(keyTypeSymbol, typeSymbol, references);
+            // add the base type
+            AddToReferenceMap(keyTypeSymbol, typeSymbol.BaseType, references);
+            if (typeSymbol is INamedTypeSymbol namedType)
+            {
+                // add the generic type arguments
+                foreach (var typeArgument in namedType.TypeArguments)
+                {
+                    AddToReferenceMap(keyTypeSymbol, typeArgument, references);
+                }
+            }
+        }
+
+        private void AddMethodSymbol(INamedTypeSymbol keySymbol, IMethodSymbol methodSymbol, Dictionary<INamedTypeSymbol, HashSet<INamedTypeSymbol>> references)
+        {
+            // add the return type
+            AddTypeSymbol(keySymbol, methodSymbol.ReturnType, references);
+            // add the parameters
+            foreach (var parameter in methodSymbol.Parameters)
+            {
+                AddTypeSymbol(keySymbol, parameter.Type, references);
+            }
+        }
+
+        private void AddToReferenceMap(INamedTypeSymbol keySymbol, ISymbol? valueSymbol, Dictionary<INamedTypeSymbol, HashSet<INamedTypeSymbol>> references)
+        {
+            // we only add the symbol when it is a named type. Anonymous type is not our concern, we just ignore them
+            if (valueSymbol is not INamedTypeSymbol valueTypeSymbol)
+                return;
+
+            references.AddInList(keySymbol, valueTypeSymbol, _symbolSetInitializer);
+        }
+
+        /// <summary>
+        /// Returns the node that defines <paramref name="node"/> inside the document under the syntax root of <paramref name="root"/>, which should be <see cref="ClassDeclarationSyntax"/>, <see cref="StructDeclarationSyntax"/> or <see cref="EnumDeclarationSyntax"/>
+        /// The <paramref name="node"/> here should come from the result of <see cref="SymbolFinder"/>, therefore a result is guaranteed
+        /// </summary>
+        /// <param name="root"></param>
+        /// <param name="node"></param>
+        /// <returns></returns>
+        private static BaseTypeDeclarationSyntax GetOwner(SyntaxNode root, SyntaxNode node)
+        {
+            var candidates = root.DescendantNodes().OfType<BaseTypeDeclarationSyntax>();
+            var result = candidates.First(candidate => candidate.DescendantNodesAndSelf().Contains(node));
+            return result;
```

Comment:
It should be much faster to go from the node up to the root.

Reply:
I did not get the point. And per my explanation in the above comment about the "documentCache", this piece of code will be reached very rarely.

Reply:
I get it now. Now we find its owner from the node up to the root and we actually do not need the root here.